rust/src/delta.rs

```diff
@@ -1355,6 +1455,223 @@ fn log_entry_from_actions(actions: &[Action]) -> Result<String, serde_json::Erro
+    DeltaLogSchema {
+        /// The source DeltaLogSchemaError
+        #[from]
+        source: DeltaLogSchemaError,
+    },
+    /// Passthrough error returned when calling DeltaTable.
+    #[error("DeltaTableError: {source}")]
+    DeltaTable {
+        /// The source DeltaTableError.
+        #[from]
+        source: DeltaTableError,
+    },
+    /// Error returned when the parquet writer fails while writing the checkpoint.
+    #[error("Failed to write parquet: {}", .source)]
+    ParquetError {
+        /// Parquet error details returned when writing the checkpoint failed.
+        #[from]
+        source: ParquetError,
+    },
+    /// Error returned when converting the schema to Arrow format failed.
+    #[error("Failed to convert into Arrow schema: {}", .source)]
+    ArrowError {
+        /// Arrow error details returned when converting the schema in Arrow format failed
+        #[from]
+        source: ArrowError,
+    },
+    /// Passthrough error returned when calling StorageBackend.
+    #[error("StorageError: {source}")]
+    Storage {
+        /// The source StorageError.
+        #[from]
+        source: StorageError,
+    },
+    /// Passthrough error returned by serde_json.
+    #[error("serde_json::Error: {source}")]
+    JSONSerialization {
+        /// The source serde_json::Error.
+        #[from]
+        source: serde_json::Error,
+    },
+}
+
+/// Struct for writing checkpoints to the delta log.
+pub struct CheckPointWriter {
+    table_uri: String,
+    delta_log_path: String,
+    last_checkpoint_path: String,
+    storage: Box<dyn StorageBackend>,
+    schema_factory: DeltaLogSchemaFactory,
+}
+
+impl CheckPointWriter {
+    /// Creates a new CheckPointWriter.
+    pub fn new(table_uri: &str, storage: Box<dyn StorageBackend>) -> Self {
+        let delta_log_path = storage.join_path(table_uri, "_delta_log");
+        let last_checkpoint_path = storage.join_path(delta_log_path.as_str(), "_last_checkpoint");
+        let schema_factory = DeltaLogSchemaFactory::new();
+
+        Self {
+            table_uri: table_uri.to_string(),
+            delta_log_path,
+            last_checkpoint_path,
+            storage,
+            schema_factory,
+        }
+    }
+
+    /// Creates a new checkpoint at the specified version.
+    /// NOTE: This method loads a new instance of delta table to determine the state to
+    /// checkpoint.
+    pub async fn create_checkpoint_for_version(
+        &self,
+        version: DeltaDataTypeVersion,
+    ) -> Result<(), CheckPointWriterError> {
+        let table = open_table_with_version(self.table_uri.as_str(), version).await?;
+
+        self.create_checkpoint_from_state(version, table.get_state())
+            .await
+    }
+
+    /// Creates a new checkpoint at the specified version from the given DeltaTableState.
+    pub async fn create_checkpoint_from_state(
+        &self,
+        version: DeltaDataTypeVersion,
+        state: &DeltaTableState,
```

Comment:
The version, state and fields in self here is basically all the fields in the `DetalTable` struct, I feel like it would be simpler if we just make this a function that takes a `DeltaTable` struct reference as input or a method of the `DeltaTable` struct. This way, we won't need to define a dedicated `CheckPointWriter` struct anymore.

Or are you thinking that there is a valid use-case where one might want to manually construct delta table states and versions without through through `DeltaTable` struct?

Reply:
I was thinking about a use case where an application might want to run data writes and checkpoints on separate threads. This was actually related to my initial design in https://github.com/delta-io/delta-rs/discussions/273 that included a channel for writing checkpoints from a separate thread in the same process while leaving the table instance available for further write actions. Theoretically, one might have a primary write thread, after every 10th commit, clone the state, send it as part of the message to the channel and keep using the same table on the primary write thread. I think we are going to use lambda checkpoints which are decoupled at the process level at this point though, so it is not as important to maintain this API. To your point though, even with the design in https://github.com/delta-io/delta-rs/discussions/273, we could clone the `DeltaTable`. Idk - given that thought process, whats your feedback?

Reply:
Ha makes sense. Now I see where you are coming from. Based on this, I am cool with this design. The idea of cloning `DeltaTable` also sounds like a good idea. I will leave that decision to you :)

Reply:
I thought about this a little more, and for public consumption, I prefer the two methods we have. We have an api that takes a version parameter, and opens a delta table at that version internally. We also have an api that takes version and state in case the caller wants to pass in memory state that they have in hand already. I like that these two APIs are very explicit about exactly what version and state will be checkpointed. An API that takes a DeltaTable would use the version and state on the DeltaTable instance so it does not communicate the content of the checkpoint that will be created as explicitly IMHO.